### ISSUE TITLE
[Feature]: @cached_property that only awaits the wrapped method _once_.

### ISSUE THREAD

**Issue by mjpieters** [maxfischer2781/asyncstdlib#149]
### Describe the feature

The `@cached_property` implementation allows for concurrent awaits on the property, with the last call to complete setting the final value:

```python
import asyncio
import random

from asyncstdlib.functools import cached_property


class Foo:
    @cached_property
    async def bar(self) -> int:
        await asyncio.sleep(random.uniform(0.01, 0.1))
        value = random.randrange(1, 100)
        print("Foo.bar ->", value)
        return value


async def demo():
    foo = Foo()
    results = await asyncio.gather(foo.bar, foo.bar, foo.bar)
    print(*results)
    print(await foo.bar)


if __name__ == "__main__":
    asyncio.run(demo())
```

The above script outputs something like:

```plain
% python demo.py
Foo.bar -> 90
Foo.bar -> 34
Foo.bar -> 76
34 76 90
76
```

I'd like to see a decorator where only the _first_ task determines the value, and any concurrent access before the first task completes having to wait for the result to have been computed. This could be achieved if the decorator first assigned an intermediary object to the instance `__dict__`, one that holds an `asyncio.Future()` instance for subsequent (concurrent) calls. Here is what that could look like (using Python 3.12 type annotation syntax for generics):

```python
class _FutureCachedValue[I, T]:
    __slots__ = ("_get_attribute", "_instance", "_fut")

    _fut: asyncio.Future[T]

    def __init__(
        self, get_attribute: Callable[[I], Coroutine[Any, Any, T]], instance: I
    ):
        self._get_attribute = get_attribute
        self._instance = instance

    def __await__(self) -> Generator[Any, Any, T]:
        try:
            fut = self._fut
        except AttributeError:
            # _fut is not yet set, this is the first accessing task
            return self._fill_future().__await__()
        else:
            # this is another, concurrent task, have it wait for the cache to be filled
            return fut.__await__()

    async def _fill_future(self) -> T:
        self._fut = asyncio.Future()
        try:
            value = await self._get_attribute(self._instance)
        except Exception as e:
            # share the exception across all waiting tasks
            self._fut.set_exception(e)
            # but allow a new task to re-attempt to fetch the value
            del self._fut
            raise
        self._fut.set_result(value)
        return value
```

 - the CachedProperty descriptor `__get__` method creates the wrapper object and stores it on the instance before returning it:
    ```
        if instance is None:
            return self
        # store a placeholder for other tasks to access the future cached value
        # on this instance.
        wrapper = _FutureCachedValue(self._get_attribute, instance)
        instance.__dict__[self._name] = wrapper
        return wrapper
    ```
     `_get_attribute` remains the same.
 - when the wrapper is awaited for the first time it will not yet have a `_fut` future, and control is passed
   to the `_fill_future()` method. This creates the `Future()` instance, then passes control to the descriptor `_get_attribute()` coroutine. Eventually, this produces a value, which is stored in the future, as well as in a `AwaitableValue()` object on the instance.
- any concurrent access to the instance _before_ a value has been produced are delegated to the future `__await__()` method and are paused.

With the above `_FutureCachedValue` used by an updated `__get__` method on `cached_property`, the output of the demo becomes:

```plain
% python demo.py
Foo.bar -> 14
14 14 14
14
```
    
### Request Assignment [Optional]

- [ ] I already have a plan and want to submit an implementation.

**Comment by mjpieters** [comment]
Note: I've updated the original report a few times as I worked out how to correctly handle concurrent access and some edge cases such as exceptions.

I've assumed that the current behaviour is intentional and the above would have to be a separate implementation from the current `cached_property` option. If it is a bug, then lets get it fixed. :-)

**Comment by maxfischer2781** [comment]
Many thanks for the report. I vaguely remember pondering this but forgetting about it. 😓 

So far this is intentional since I don't want to tie the implementation to a specific `async` flavour (or a dependency). However, similar thoughts to #86 should apply - supplying a lock (type) could be a simple compromise.
```
class Foo:
    @cached_property  # vvv may complete multiple times before the value is cached
    async def bar(self) -> int: ...

    @cached_property(asyncio.Lock)  # vvv completes only once before the value is cached
    async def qux(self) -> int: ...
```
(Give or take a required parameter name that makes the synchronisation effect of the lock apparent.)

Regardless whether this is a separate utility or not, I would prefer a lock-based implementation since those play nicer with retrying when the worker is cancelled. But the general structure should indeed be as you describe it.

**Comment by mjpieters** [comment]
> So far this is intentional since I don't want to tie the implementation to a specific `async` flavour (or a dependency).

Ah, yes, sorry! I forgot about that.

Yes, accepting an arbitrary async context manager would let you remain agnostic here, and would let you get creative with different synchronisation primitives. I now have to go find a use-case for using a barrier for this 🙃 

So `descriptor.__get__` then produces a placeholder object to attach the lock to an instance. When awaited it will get the attribute from the instance, if it is the placeholder object it attempts to take the lock. Once it has the lock it checks again if the instance still has the placeholder object, and only then runs the method. Other tasks holding the placeholder either await on it after the value was cached and so get the cached value, or wait for the lock, _then_ find the cached value. Sounds solid.

This does change how exceptions in the getter are handled; they are only propagated to the task that holds the lock, the other tasks re-run the getter.

A nice side effect is that if the placeholder is deleted from the instance, then any task still holding that placeholder object will automatically restart the process.

Here's my sketch version of the placeholder:

```python
class _FutureCachedValue[I, T]:
    """A placeholder object to control concurrent access to a cached awaitable value.

    When given a lock to coordinate access, only the first task to await on a
    cached property triggers the underlying coroutine. Once a value has been
    produced, all tasks are unblocked and given the same, single value.

    """

    __slots__ = ("_get_attribute", "_instance", "_name", "_cm")

    def __init__(
        self,
        get_attribute: Callable[[S], Coroutine[Any, Any, T]],
        instance: S,
        name: str,
        cm: AsyncContextManager[Any],
    ):
        self._get_attribute = get_attribute
        self._instance = instance
        self._name = name
        self._cm = cm

    def __await__(self) -> Generator[None, None, T]:
        return self._await_impl().__await__()

    async def _await_impl(self) -> T:
        try:
            stored = t.cast(Awaitable[T], self._instance.__dict__[self._name])
        except KeyError:
            # something deleted the cached value or future cached value placeholder. Restart
            # the fetch by delegating to the cached_property descriptor.
            return await t.cast(Awaitable[T], getattr(self._instance, self._name))

        if stored is not self:
            # another task produced a value
            return await stored

        # attempt to get the lock
        async with self._lock:
            # check again for a cached value
            try:
                stored = t.cast(Awaitable[T], self._instance.__dict__[self._name])
            except KeyError:
                # something deleted the cached value or future cached value placeholder. Restart
                # the fetch by delegating to the cached_property descriptor.
                return await t.cast(Awaitable[T], getattr(self._instance, self._name))

            if stored is not self:
                # another task produced a value
                return await stored

            # the instance attribute is still this placeholder, and we hold the lock. Start the getter
            # to store the value on the instance and return the value.
            return await self._get_attribute(self._instance)
```

Here is an updated demonstrator script, which assumes there is an updated `cached_property` object that accepts an async contextmanager type as the lock type and returns an instance of the above placeholder on `__get__()` after storing that placeholder on the instance:

```python
import asyncio
import random
import sys
import typing as t

from asyncstdlib import cached_property


class Foo:
    async def bar(self) -> int:
        await asyncio.sleep(random.uniform(0.01, 0.1))
        return random.randrange(1, 100)

    @cached_property
    async def unlocked(self) -> int:
        value = await self.bar()
        print("  Foo.unlocked call ->", value)
        return value

    @cached_property(asyncio.Lock)
    async def locked(self) -> int:
        value = await self.bar()
        print("  Foo.locked call ->", value)
        return value


async def demo():
    foo = Foo()
    print("Stored attributes:", foo.unlocked, foo.locked)

    async def task(attr: t.Awaitable[int]):
        return await attr

    print("Unlocked:")
    results = await asyncio.gather(
        task(foo.unlocked), task(foo.unlocked), task(foo.unlocked)
    )
    print("Task results:", *results)
    print("Final stored value:", await foo.unlocked)

    print("Locked:")
    results = await asyncio.gather(task(foo.locked), task(foo.locked), task(foo.locked))
    print("Task results:", *results)
    print("Final stored value:", await foo.locked)

    print("Stored attributes:", foo.unlocked, foo.locked)


if __name__ == "__main__":
    asyncio.run(demo())
```

The output looks like:

```plain
Stored attributes: <util.asyncio._FutureCachedValue object at 0x103252380> <util.asyncio._FutureCachedValue object at 0x103252340>
Unlocked:
  Foo.unlocked call -> 48
  Foo.unlocked call -> 33
  Foo.unlocked call -> 3
Task results: 33 3 48
Final stored value: 3
Locked:
  Foo.locked call -> 3
Task results: 3 3 3
Final stored value: 3
Stored attributes: AwaitableValue(3) AwaitableValue(3)
```
